### DISCUSSION TITLE
lens conjunction doesn't work

### DISCUSSION THREAD

**Discussion by LLFourn** [yamafaktory/jql#171] [General]
I would expect the following to output only the second element:

```
printf '[{"cond1": "false", "cond2": "true"}, {"cond1": "true", "cond2": "true"}]' | jql '.[]|={"cond1":"true"}|={"cond2":"true"}'
```
but it outputs
```json
[
  {
    "cond1": "false",
    "cond2": "true"
  },
  {
    "cond1": "true",
    "cond2": "true"
  }
]
```

How do you do a conjunction on the lens conditions?

**Comment by yamafaktory** [comment]
This is not a bug, lenses - by design - were not made for that purpose. It's more about filtering the data so e.g. given the following `test.json` file:

```json
[
  {
    "cond1": "false",
    "cond2": "true",
    "foo": 123
  },
  {
    "cond1": "true",
    "cond2": "true",
    "bar": 456
  },
  {
    "cond1": "true",
    "cond2": "false",
    "yup": 789
  }
]
```

You can pick two elements that are holding some key/value pairs:

```sh
 jql '.|={"cond1":"false","yup":"789"}' test.json
```
```json
[
  {
    "cond1": "false",
    "cond2": "true",
    "foo": 123
  },
  {
    "cond1": "true",
    "cond2": "false",
    "yup": 789
  }
]
```
Moving this to a discussion since changing this would break the actual grammar and the goal was to keep the tool as simple as possible.

**Reply by LLFourn** [reply]
Right. I understood that the syntax you used was a disjunction bwetween each condition but I expected `|={COND1}|={COND2}` to apply `COND2` on the set of elements that passed the `COND1` lens. I mean the pipe symbol seems to be hinting at POSIX shell pipe semantics but in this case it's not.  For example you could do what I want with two invocations of jql piped together.

```
printf '[{"cond1": "false", "cond2": "true"}, {"cond1": "true", "cond2": "true"}]' | jql '.[]|={"cond1":"true"}' | jql '.[]|={"cond2":"true"}'
```

It is quite unintuitive that:

```
jql '.[]|={"cond1":"true"}' | jql '.[]|={"cond2":"true"}
```
is not the same as a single
```
jql '.[]|={"cond1":"true"}|={"cond2":"true"}' 
```

**Reply by yamafaktory** [reply]
Gotcha, thanks for clarifying. Indeed something needs to be fixed, I will have a look at it.

**Reply by yamafaktory** [reply]
@LLFourn I've just release a new major version that fixes the bug:

- lenses and default filters are now merged into one enum (hence the major version since some other projects are consuming `jql` as a library)
- the pest grammar has be simplified
- piping lenses together - or with default filters - will just work as expected

Thanks a ton for submitting the initial issue and for pushing forward 🚀, you've helped me to fix a major flow in the tool!

**Reply by LLFourn** [reply]
Fantastic work. Cheers mate.
